touchforms/backend/touchcare.py

```diff
@@ -31,264 +34,53 @@
-
-    def load_object_ids(self):
-        if self.form_context.get('all_case_ids', None):
-            case_ids = self.form_context.get('all_case_ids')
-        else:
-            case_ids = query_case_ids(self.query_func, criteria=self.additional_filters)
-        # todo: see note above about why sorting is necessary
-        self.ids = dict(enumerate(sorted(case_ids)))
-
-    def getIDsForValue(self, field_name, value):
-        logger.debug('case index lookup %s %s' % (field_name, value))
-
-        if (field_name, value) not in self.cached_lookups:
-            if field_name == 'case-id':
-                cases = [self.read_object(value)]
-            else:
-                try:
-                    get_val = {
-                        'case-type': lambda c: c.getTypeId(),
-                        'case-status': lambda c: 'closed' if c.isClosed() else 'open',
-                    }[field_name]
-                except KeyError:
-                    # Try any unrecognized field name as a case id field.
-                    # Needed for 'case-in-goal' lookup in PACT Care Plan app.
-                    cases = [self.read_object(value)]
-                else:
-                    cases = [c for c in self.objects.values() if get_val(c) == value]
-
-            self.cached_lookups[(field_name, value)] = cases
-
-        cases = self.cached_lookups[(field_name, value)]
-        id_map = dict((v, k) for k, v in self.ids.iteritems())
+    def needs_sync(self):
         try:
-            return to_vect(id_map[c.getCaseId()] for c in cases)
-        except KeyError:
-            # Case was not found in id_map
-            raise CaseNotFound
-
-
-class LedgerDatabase(TouchformsStorageUtility):
-    def __init__(self, host, domain, auth, additional_filters, preload):
-        super(LedgerDatabase, self).__init__(host, domain, auth, additional_filters, preload)
-
-    def get_object_id(self, ledger):
-        return ledger.getEntiyId()
-
-    def fetch_object(self, entity_id):
-        return query_ledger_for_case(self.query_func, entity_id)
-
-    def load_object_ids(self):
-        case_ids = query_case_ids(self.query_func, criteria=self.additional_filters)
-        self.ids = dict(enumerate(case_ids))
-
-    def getIDsForValue(self, field_name, value):
-        logger.debug('ledger lookup %s %s' % (field_name, value))
-        if (field_name, value) not in self.cached_lookups:
-            if field_name == 'entity-id':
-                ledgers = [self.read_object(value)]
-            else:
-                raise NotImplementedError("Only entity-id lookup is currently supported!")
-
-            self.cached_lookups[(field_name, value)] = ledgers
-
-        else:
-            ledgers = self.cached_lookups[(field_name, value)]
-
-        id_map = dict((v, k) for k, v in self.ids.iteritems())
-        return to_vect(id_map[l.getEntiyId()] for l in ledgers)
+            self.last_sync = self.sandbox.getLastSync()
+        except:
```

Comment:
My thinking was basically that we probably never want to crash if we can't get the last sync, since its easy just to clear and restore. Valid? I'll add logging for sure